core/kotlinx-coroutines-core/src/internal/CustomServiceLoader.kt

```diff
@@ -0,0 +1,116 @@
+package kotlinx.coroutines.internal
+
+import java.io.*
+import java.lang.ClassCastException
+import java.net.URL
+import java.util.*
+import java.util.jar.JarFile
+import java.util.zip.ZipEntry
+import kotlin.NoSuchElementException
+import java.io.IOException
+import java.io.BufferedReader
+
+private const val PREFIX: String = "META-INF/services/"
+
+public fun <S> loadProviders(service: Class<S>, loader: ClassLoader) = CustomServiceLoader(service, loader)
+
+public class CustomServiceLoader<S> internal constructor (private val service: Class<S>, private val loader: ClassLoader?) : Iterable<S> {
+
+    private fun parse(service: Class<S>, url: URL): Iterator<String> {
+        val names = mutableListOf<String>()
+        try {
+            val string = url.toString()
+            val separatorIndex = string.indexOf('!')
+            val pathToJar = string.substring(0, separatorIndex)
+            val entry = string.substring(separatorIndex + 2, string.length)
+            (JarFile(pathToJar.substring("jar:file:/".length), false) as Closeable).use { file ->
```

Comment:
It looks like ad-hoc checks. Please leave some comments about what is happening here and why it is legal to unconditionally assume prefixes and `!` symbols